File: .API/ComputationLock.cs
// Decompiled with JetBrains decompiler
// Type: CloudX.Shared.ComputationLock
// Assembly: CloudX.Shared, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null
// MVID: 2635624C-5F24-4EFB-ACD1-7E9C1349B2F5
// Assembly location: F:\SteamLibrary\steamapps\common\NeosVR\HeadlessClient\CloudX.Shared.dll

using Newtonsoft.Json;
using System;

namespace CloudX.Shared
{
  [JsonObject(MemberSerialization = MemberSerialization.OptIn)]
  public struct ComputationLock
  {
    [JsonProperty(PropertyName = "token")]
    public string Token { get; private set; }

    [JsonProperty(PropertyName = "timestamp")]
    public DateTime ExpireTimestamp { get; private set; }

    public bool IsLocked
    {
      get
      {
        if (string.IsNullOrWhiteSpace(this.Token))
          return false;
        return DateTime.UtcNow > this.ExpireTimestamp;
      }
    }

    public bool TryLock(TimeSpan duration)
    {
      if (this.IsLocked)
        return false;
      this.Token = Guid.NewGuid().ToString();
      this.ExpireTimestamp = DateTime.UtcNow + duration;
      return true;
    }

    public bool TryExtend(string token, TimeSpan duration)
    {
      if (token != this.Token)
        return false;
      this.ExpireTimestamp = DateTime.UtcNow + duration;
      return true;
    }

    public bool TryRelease(string token)
    {
      if (this.Token != token)
        return false;
      this.Token = (string) null;
      this.ExpireTimestamp = new DateTime();
      return true;
    }
  }
}

File: src/api/ComputationLock.ts
import { TimeSpan } from "@bombitmanbomb/utils";
import { v4 as uuidv4 } from "uuid";
export class ComputationLock {
	public Token: string;
	public ExpireTimestamp: Date;
	constructor($b: ComputationLockJSON) {
		this.Token = $b.token;
		this.ExpireTimestamp = $b.timestamp;
	}
	public get IsLocked(): boolean {
		return (
			!(this.Token == null || this.Token.trim() == "") &&
			new Date() > this.ExpireTimestamp
		);
	}
	public TryLock(duration: TimeSpan): boolean {
		if (this.IsLocked) return false;
		this.Token = uuidv4();
		this.ExpireTimestamp = new Date(new Date().getTime() + duration.msecs);
		return true;
	}
	public TryExtend(token: string, duration: TimeSpan): boolean {
		if (token != this.Token) return false;
		this.ExpireTimestamp = new Date(new Date().getTime() + duration.msecs);
		return true;
	}
	public TryRelease(token: string): boolean {
		if (this.Token != token) return false;
		this.Token = "";
		this.ExpireTimestamp = new Date();
		return true;
	}
	toJSON(): ComputationLockJSON {
		return { token: this.Token, timestamp: this.ExpireTimestamp };
	}
}
export interface ComputationLockJSON {
	token: string;
	timestamp: Date;
}
